### ISSUE TITLE
request: json_init_library

### ISSUE THREAD

**Issue by dorssel** [json-c/json-c#540]
Many libraries allow overriding `malloc`, `realloc`, and `free` by (optionally) calling an initialization function. One of the use cases is to securely zeroize sensitive buffers. Currently, json-c uses platform `malloc` and `free`, which leaves all kinds of data (think: passwords) on the heap.

I suggest to use new functions `json_malloc` etc. throughout the library, which default to `malloc` etc., but which can be set to user provided functions (only upon initialization). Currently, json-c also uses  `strdup`, which could be replaced by an internal function that implements it using the new `json_malloc`.

**Comment by hawicz** [comment]
If we're going to go through the trouble of changing every memory allocation, it would be worthwhile to also use a type-specific slab allocators for the struct json_object's, struct array_list's, struct lh_table's, etc...  Then again, that's a rather more significant change.

As for zeroizing sensitive buffers, if you're worried about that then it seems that merely changing the allocator used by json-c would be insufficient, as you'd be likely to have other allocations in your program that might have worthwhile-to-be-wiped content, in which case you should just replace malloc, free, etc... for the entire process.

In short, I'm not actually convinced that overriding library-specific memory allocation in this way is worthwhile.

**Comment by dota17** [comment]
@hawicz As you said, slab allocator is acturaly suitable for the fixed-size struct/object  which is scheduled frequently. It is a good optimization direction. I will try to implement this function 0.0. 
Using malloc/free will casue sensitive data leak. And malloc is not good func which also brings memory leak. In short, rewrite malloc/free is complicated and we shouldconsider it carefully.

**Comment by hawicz** [comment]
@dota17, before implementing anything a change like that is worth discussing in more detail.  If you have some ideas about an existing allocator to bring in to json-c, please mention it on the json-c Google group, or in a separate github issue.

**Comment by dota17** [comment]
@hawicz  ok, I got it!

**Comment by hawicz** [comment]
I don't think there's anything to be done for this, so I'm closing it.
